src/merkle_tree/incremental_merkle_tree.rs

```diff
@@ -0,0 +1,277 @@
+use crate::crh::TwoToOneCRHScheme;
+use crate::merkle_tree::{tree_height, Config, DigestConverter, LeafParam, Path, TwoToOneParam};
+use crate::CRHScheme;
+use ark_std::borrow::Borrow;
+
+/// Defines an incremental merkle tree data structure.
+/// This merkle tree has runtime fixed height, and assumes number of leaves is 2^height.
+///
+#[derive(Derivative)]
+#[derivative(Clone(bound = "P: Config"))]
+pub struct IncrementalMerkleTree<P: Config> {
+    /// Store the hash of leaf nodes from left to right
+    leaf_nodes: Vec<P::LeafDigest>,
+    /// Store the inner hash parameters
+    two_to_one_hash_param: TwoToOneParam<P>,
+    /// Store the leaf hash parameters
+    leaf_hash_param: LeafParam<P>,
+    /// Stores the height of the MerkleTree
+    height: usize,
+    /// Stores the path of the "current leaf"
+    /// if current_path.auth_path.len() == mean, means the IMT is empty
```

Comment:
I think this comment line should be removed.

### Explanation

I think by `== mean` you meant `== 0`. Even so, this is not true, since a Merkle Tree of height 2 (capacity 2) with size 1 (one true leaf and one blank leaf) has an empty `current_path.auth_path`. The only way to check if the tree is empty at the moment is to use the empty flag as used or to use one of the following weaker alternatives:

1. Check the root is default (fails if there exists a preimage attack)
```rust
self.root == Default::default()
```

2. Check the root is default and that the proof is empty (fails if there exists a preimage attack and the tree height is 2).
```rust
self.root == Default::default() && self.current_path.auth_path.is_empty()
```

You could also use `self.leaf_nodes.is_empty()` but this also fails for trees of size 1.

Reply:
good call. thanks!

Reply:
I think for the exact reason I add a boolean field empty. Forget to update the comment.